pennylane/data/base/dataset.py

```diff
@@ -0,0 +1,361 @@
+            attributes = self.attrs
+
+        with h5py.open_group(filepath, mode=mode) as grp:
+            grp.attrs.update(self.bind.attrs)
+            for attr in attributes:
+                h5py.copy(self.bind[attr], grp, attr)
+
+    def _validate_attrs(self) -> None:
+        """Validates that ``attrs`` matched the delcared fields of this
+        dataset."""
+        missing = []
+        for name in self.fields:
+            if name not in self.attrs:
+                missing.append(name)
+
+        if missing:
+            missing_args = ", ".join(f"'{arg}'" for arg in missing)
+            raise TypeError(
+                f"__init__() missing {len(missing)} required keyword argument(s): {missing_args}"
+            )
+
+    def _init_bind(self):
+        if self.bind.file.mode == "r+":
+            if "type_id" not in self.info:
+                self.info["type_id"] = self.type_id
+            if "category_id" not in self.info and (getattr(self, "category_id", None) is not None):
+                self.info["category_id"] = self.category_id
+
+    def __setattr__(self, __name: str, __value: Union[Any, AttributeType]) -> None:
+        if __name.startswith("_") or __name in type(self).__dict__:
+            object.__setattr__(self, __name, __value)
+            return
+
+        if __name in self.fields:
+            field = self.fields[__name]
+            self._mapper.set_item(__name, __value, field.info, field.attribute_type)
+        else:
+            self._mapper[__name] = __value
+
+    def __getattr__(self, __name: str) -> Any:
+        try:
+            return self._mapper[__name].get_value()
+        except KeyError as exc:
+            if __name in self.fields:
+                return UNSET
+
+            raise AttributeError(f"'{type(self)}' object has no attribute '{__name}'") from exc
+
+    def __delattr__(self, __name: str) -> None:
+        try:
+            del self._mapper[__name]
+        except KeyError as exc:
+            raise AttributeError(f"'{type(self)}' object has no attribute '{__name}'") from exc
+
+    def __repr__(self) -> str:
+        repr_items = ", ".join(
+            f"{name}: {value}"
+            for name, value in {**self.params, "attributes": (str(list(self.attrs)))}.items()
+        )
+
+        return f"<{type(self).__name__} = {repr_items}>"
+
+    def __init_subclass__(cls) -> None:
+        """Initializes the ``fields`` dict of a Dataset subclass using
+        the declared ``Attributes`` and their type annotations."""
+
+        super().__init_subclass__()
+
+        fields = {}
+        reserved_names = ("params", "bind", "validate")
+
+        for name, annotated_type in cls.__annotations__.items():
+            if (
+                name in reserved_names
+                or isinstance(annotated_type, InitVar)
+                or get_origin(annotated_type) is ClassVar
+            ):
+                continue
+
+            attr = getattr(cls, name, attribute())
+            try:
```

Comment:
we could probably avoid this try-except based on the result of the `getattr` above, and we try (hehe) to avoid them as much as possible

Reply:
i meant to avoid the try statement altogether, but `__init_subclass__` shouldn't be called that many times anyway so it's all good.